cpp/src/arrow/filesystem/gcsfs.h

```diff
@@ -77,6 +83,13 @@ struct ARROW_EXPORT GcsOptions {
   /// This will be ignored if non-empty metadata is passed to OpenOutputStream.
   std::shared_ptr<const KeyValueMetadata> default_metadata;
 
+  /// \brief The project to use for creating buckets.
+  ///
+  /// If not set, the library uses the GOOGLE_CLOUD_PROJECT environment
+  /// variable. Most I/O operations do not need a project id, only applications
+  /// that create new buckets need a project id.
+  std::optional<std::string> project_id;
```

Comment:
FTR, an optional was probably not necessary as I suppose an empty string cannot possibly be a valid project id?

Reply:
You are probably right.  There is a difference between "not set" (which implies "use the `GOOGLE_CLOUD_PROJECT` environment variable") and "set to empty" (which implies "ignore the environment variable and use the empty string"). Is this distinction useful?  Arguably not.

I am working on the Python fixes today, I can change this to `std::string` if you think it is best to do so.

Reply:
It's probably ok to keep the optional.
There's already a PR for Python here: https://github.com/apache/arrow/pull/36376

Reply:
hey @coryan I gave the Python PR a try to try and fix some of the CI failures. I am trying to fix some of our nightly builds to try and get to a better state before the code freeze. Sorry for stepping in :)

Reply:
No worries at all, I am sure your fixes will be better.  And I am getting very efficient at "did nothing" 😄